internal/backend/runtime/omni/controllers/omni/talos_upgrade_status_test.go

```diff
@@ -6,333 +6,523 @@
+				func(res *omni.ClusterMachine) error {
+					res.TypedSpec().Value.KubernetesVersion = cluster.TypedSpec().Value.KubernetesVersion
+
+					return nil
+				},
+			),
+		),
+	)
 
-	clusterName := "talos-upgrade-locked"
+	workerMachines := rmock.MockList[*omni.ClusterMachine](ctx, t, st,
+		testoptions.QueryIDs[*omni.MachineSetNode](resource.LabelEqual(omni.LabelMachineSet, workersMachineSet.Metadata().ID())),
+		testoptions.ItemOptions(
+			testoptions.LabelCluster(cluster),
+			testoptions.LabelMachineSet(workersMachineSet),
+			testoptions.Modify(
+				func(res *omni.ClusterMachine) error {
+					res.TypedSpec().Value.KubernetesVersion = cluster.TypedSpec().Value.KubernetesVersion
+
+					return nil
+				},
+			),
+		),
+	)
 
-	cluster, machines := suite.createCluster(clusterName, 1, 3)
+	machines := slices.Concat(cpMachines, workerMachines)
 
-	clusterStatus := omni.NewClusterStatus(resources.DefaultNamespace, clusterName)
-	clusterStatus.TypedSpec().Value.Available = true
-	clusterStatus.TypedSpec().Value.Ready = true
-	clusterStatus.TypedSpec().Value.Phase = specs.ClusterStatusSpec_RUNNING
+	for _, machine := range machines {
+		rmock.Mock[*siderolink.MachineJoinConfig](ctx, t, st, testoptions.SameID(machine),
+			testoptions.Modify(func(res *siderolink.MachineJoinConfig) error {
+				res.TypedSpec().Value.Config = &specs.JoinConfig{}
 
-	suite.Require().NoError(suite.state.Create(suite.ctx, clusterStatus))
+				return nil
+			}),
+		)
 
-	for _, res := range machines {
-		if _, ok := res.Metadata().Labels().Get(omni.LabelWorkerRole); ok {
-			_, err := safe.StateUpdateWithConflicts(suite.ctx, suite.state, resource.NewMetadata(
-				resources.DefaultNamespace, omni.MachineSetNodeType, res.Metadata().ID(), resource.VersionUndefined,
-			), func(r *omni.MachineSetNode,
-			) error {
-				r.Metadata().Annotations().Set(omni.MachineLocked, "")
+		rmock.Mock[*siderolink.Link](ctx, t, st, testoptions.SameID(machine),
+			testoptions.Modify(func(res *siderolink.Link) error {
+				res.TypedSpec().Value.Connected = true
 
 				return nil
-			})
+			}),
+		)
 
-			suite.Require().NoError(err)
-		}
+		rmock.Mock[*omni.MachineStatus](ctx, t, st, testoptions.SameID(machine),
+			testoptions.Modify(func(res *omni.MachineStatus) error {
+				res.TypedSpec().Value.Maintenance = false
 
-		cmtv := omni.NewClusterMachineTalosVersion(resources.DefaultNamespace, res.Metadata().ID())
+				return nil
+			}),
+		)
 
-		helpers.CopyAllLabels(res, cmtv)
+		rmock.Mock[*omni.Machine](ctx, t, st, testoptions.SameID(machine))
+		rmock.Mock[*omni.ClusterMachineStatus](ctx, t, st, testoptions.SameID(machine),
+			testoptions.Modify(func(res *omni.ClusterMachineStatus) error {
+				service := machineServices.Create(ctx, res.Metadata().ID())
+				helpers.CopyLabels(machine, res, omni.LabelCluster, omni.LabelMachineSet)
+				res.Metadata().Labels().Set(omni.LabelHostname, res.Metadata().ID())
+				res.TypedSpec().Value.ManagementAddress = service.SocketConnectionString
```

Comment:
I'll suggest to copy the management address from the machine status instead.
We can use `addDefault` in the `rmock.go`:
```go
		machineStatus, err := safe.ReaderGetByID[*omni.MachineStatus](ctx, st, res.Metadata().ID())
		if err != nil && !state.IsNotFoundError(err) {
			return err
		}

		if machineStatus != nil {
			res.TypedSpec().Value.ManagementAddress = machineStatus.TypedSpec().Value.ManagementAddress
		}
```